### PR TITLE
Switch to topkg

### PR DESCRIPTION
**Pull request by samoht**
Initially I wanted to fix:

```
Error: Files /Users/thomas/.opam/datakit/lib/osx-cf/cf_lwt.cmxa
       and /Users/thomas/.opam/datakit/lib/osx-cf/cf.cmxa
       both define a module named Cf_types
```

But it was simpler to just use topkg. Almost nothing to do, all the magic myocamlbuild runes were already there.

/cc @dbuenzli

### CODE REVIEW

File: opam [hunk 1]
```diff
@@ -6,16 +6,12 @@ authors: ["David Sheets" "Thomas Gazagnaire"]
...
-  [make "build"]
-]
-install: [
-  [make "install"]
-]
-remove: [make "uninstall"]
+build: ["ocaml" "pkg/pkg.ml" "build" "--with-lwt" lwt:installed]
 depends: [
   "ocamlfind" {build}
+  "topkg"     {>= "0.7.2"}
```
Comment (yallop) [inline opam]: Should `topkg` be a `{build}` dependency?

Comment (samoht) [inline opam]: thanks, fixed.

**Comment by samoht** [comment]
I am still trying to test this, so please don't merge yet.

**Comment by samoht** [comment]
Seems to work fine so far with my limited testing.

File: myocamlbuild.ml [hunk 2]
```diff
@@ -1,10 +1,15 @@
 open Ocamlbuild_plugin;;
 open Ocamlbuild_pack;;
 
-let ctypes_libdir = Sys.getenv "CTYPES_LIB_DIR" in
-let ocaml_libdir = Sys.getenv "OCAML_LIB_DIR" in
+let ctypes_libdir =
+  run_and_read ("ocamlfind query ctypes")
```
Comment (dbuenzli) [inline myocamlbuild.ml:5]: This may be problematic in cross compilation scenarios.

File: myocamlbuild.ml [hunk 3]
```diff
@@ -1,10 +1,15 @@
...
+  |> String.trim
 
-dispatch begin
+let ocaml_libdir =
+  run_and_read ("ocamlc -where")
```
Comment (dbuenzli) [inline myocamlbuild.ml]: This aswell.

File: opam [hunk 4]
```diff
@@ -6,16 +6,12 @@ authors: ["David Sheets" "Thomas Gazagnaire"]
...
+doc: "https://dsheets.github.io/ocaml-osx-cf/"
 license: "ISC"
-build: [
-  [make "build"]
-]
-install: [
-  [make "install"]
-]
-remove: [make "uninstall"]
+build: ["ocaml" "pkg/pkg.ml" "build" "--with-lwt" lwt:installed]
```
Comment (dbuenzli) [inline opam]: you are missing the `--pinned pinned` option.

File: myocamlbuild.ml (hunk 2 above)
Comment (samoht) [inline myocamlbuild.ml:5]: Can I use `OS.Cmd` in `myocamlbuild.ml`?

Comment (dbuenzli) [inline myocamlbuild.ml:5]: In fact that one will be fine with the OCAMLFIND_TOOLCHAIN hack.

File: myocamlbuild.ml (hunk 3 above)
Comment (dbuenzli) [inline myocamlbuild.ml]: So here do a `ocamlfind ocamlc -where`.

**Comment by samoht** [comment]
The CI doesn't seem to like that very much:

```
Undefined symbols for architecture x86_64:
#   "_ffi_closure_alloc", referenced from:
#       _ctypes_make_function_pointer in libctypes-foreign-base_stubs.a(ffi_call_stubs.o)
#   "_ffi_prep_closure_loc", referenced from:
#       _ctypes_make_function_pointer in libctypes-foreign-base_stubs.a(ffi_call_stubs.o)
# ld: symbol(s) not found for architecture x86_64
# clang: error: linker command failed with exit code 1 (use -v to see invocation)
# File "caml_startup", line 1:
# Error: Error during linking
# Command exited with code 2.
### stderr ###
# pkg.ml: [ERROR] cmd ['ocamlbuild' '-use-ocamlfind' '-classic-display' '-tag' 'debug'
#      '-build-dir' '_build' 'opam' 'pkg/META' 'CHANGES.md' 'LICENSE.md'
#      'README.md' 'lib/osx-cf.a' 'lib/osx-cf.cmxs' 'lib/osx-cf.cmxa'
#      'lib/osx-cf.cma' 'lib/cf_generated.cmx' 'lib/cf_bindings.cmx'
#      'lib/cf_types_detected.cmx' 'lib/cf_types.cmx' 'lib/cf.cmx' 'lib/cf.cmi'
#      'lib/cf.mli' 'lwt/osx-cf-lwt.a' 'lwt/osx-cf-lwt.cmxs'
#      'lwt/osx-cf-lwt.cmxa' 'lwt/osx-cf-lwt.cma' 'lwt/cf_lwt.cmx'
#      'lwt/cf_lwt.cmi' 'lwt/cf_lwt.mli' 'lib/dllcf_stubs.so'
#      'lib/libcf_stubs.a' 'lib_test/test.native']: exited with 10
```

(in https://travis-ci.org/samoht/irmin-watcher/jobs/149724871#L984)

**Comment by yallop** [comment]
That error looks [familiar](https://github.com/ocamllabs/ocaml-ctypes/issues/74).

**Comment by dsheets** [comment]
It's the same as the lwt/ocaml link order bug for which I have that `discover.ml` patch.

**Comment by samoht** [comment]
@dsheets so your patched version of lwt will fix that issue too?

**Comment by dsheets** [comment]
I believe it's the same issue... the ancient system-bundled version of libffi is being picked up and it lacks those symbols.

**Comment by samoht** [comment]
If I add `-lcf_stubs` when compiling the tests it works, but I now get:

```
#=== ERROR while installing osx-fsevents.0.1.0 ================================#
# opam-version 1.2.2
# os           darwin
# command      make build
# path         /Users/travis/.opam/system/build/osx-fsevents.0.1.0
# compiler     system (4.03.0)
# exit-code    2
# env-file     /Users/travis/.opam/system/build/osx-fsevents.0.1.0/osx-fsevents-9122-fbcac7.env
# stdout-file  /Users/travis/.opam/system/build/osx-fsevents.0.1.0/osx-fsevents-9122-fbcac7.out
# stderr-file  /Users/travis/.opam/system/build/osx-fsevents.0.1.0/osx-fsevents-9122-fbcac7.err
### stdout ###
# lib_gen/fsevents_types_detect.c:74:70: error: use of undeclared identifier 'kFSEventStreamEventFlagItemIsHardlink'
# [...]
# lib_gen/fsevents_types_detect.c:75:35: error: use of undeclared identifier 'kFSEventStreamEventFlagItemIsHardlink'; did you mean 'kFSEventStreamEventFlagItemIsSymlink'?
#      FSEventStreamEventFlags v = (kFSEventStreamEventFlagItemIsHardlink);
#                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#                                   kFSEventStreamEventFlagItemIsSymlink
# /System/Library/Frameworks/CoreServices.framework/Frameworks/CarbonCore.framework/Headers/FSEvents.h:401:3: note: 'kFSEventStreamEventFlagItemIsSymlink' declared here
#   kFSEventStreamEventFlagItemIsSymlink = 0x00040000,
#   ^
# 4 errors generated.
# Command exited with code 1.
```

I will try with a patched version of lwt too.

**Comment by yallop** [comment]
A link to the `discover.ml` patch, for cross-referencing: https://github.com/ocsigen/lwt/pull/215

**Comment by dsheets** [comment]
@samoht are you testing on OS X 10.10 or later?

**Comment by samoht** [comment]
Whatever version travisCI is using

**Comment by dsheets** [comment]
It depends on how you have it configured. I think you need to set your XCode version to 7.3 or later or something to pick up 10.10 or 10.11. The particular missing symbol arrived in 10.10.

**Comment by samoht** [comment]
I confirm than using @dsheets's patched lwt make everything works: https://github.com/samoht/irmin-watcher/pull/1

**Comment by dsheets** [comment]
Great! Thanks!

**Pull request merged**